File: ios/OakApp/Features/Teams/TeamsAssistantSheet.swift
import SwiftUI

/// The team-builder assistant, presented as a sheet from the team editor's toolbar —
/// iOS's native stand-in for web's docked `TeamsAssistantPanel.tsx`. On a phone a
/// sheet (rather than a side rail) keeps the editor form full-width and gives the
/// assistant a focused, dismissible surface; the LIVE unsaved draft rides every turn
/// regardless of which surface is on top.
///
/// A deliberately lean chat surface (NOT the full `AnswerCard` tree): user bubbles,
/// streamed assistant Markdown, a tool-activity ticker, and — when an answer proposes a
/// ``TeamPatch`` — a "Proposed changes" card with Apply / Undo. Apply mutates the
/// editor's in-memory draft only (``TeamsAssistantViewModel/apply(_:)``); the user still
/// reviews and Saves. All state lives in the injected ``TeamsAssistantViewModel``.
struct TeamsAssistantSheet: View {
  @State private var model: TeamsAssistantViewModel
  @State private var input = ""
  @Environment(\.dismiss) private var dismiss

  init(model: TeamsAssistantViewModel) {
    _model = State(initialValue: model)
  }

  var body: some View {
    NavigationStack {
      VStack(spacing: 0) {
        thread
        Divider()
        composer
      }
      .navigationTitle("Team assistant")
      .navigationBarTitleDisplayMode(.inline)
      .toolbar {
        ToolbarItem(placement: .topBarTrailing) {
          Button("Done") { dismiss() }
        }
      }
      .onDisappear { model.cancel() }
    }
  }

  // MARK: Thread

  private var thread: some View {
    ScrollViewReader { proxy in
      ScrollView {
        LazyVStack(alignment: .leading, spacing: 16) {
          if model.turns.isEmpty && model.status != .thinking {
            intro
          }

          ForEach(model.turns) { turn in
            turnView(turn)
              .id(turn.id)
          }

          if model.status == .thinking {
            pending
              .id(pendingAnchor)
          }

          if let errorMessage = model.errorMessage {
            ErrorBanner(
              message: errorMessage,
              retryTitle: "Retry",
              onRetry: { model.retry() }
            )
            .id(errorAnchor)
          }
        }
        .padding(16)
      }
      .onChange(of: model.turns.count) { _, _ in scrollToEnd(proxy) }
      .onChange(of: model.streamingMarkdown) { _, _ in scrollToEnd(proxy) }
      .onChange(of: model.activity) { _, _ in scrollToEnd(proxy) }
      .onChange(of: model.errorMessage) { _, _ in scrollToEnd(proxy) }
    }
  }

  private let pendingAnchor = -1
  private let errorAnchor = -2

  private func scrollToEnd(_ proxy: ScrollViewProxy) {
    let target: Int
    if model.errorMessage != nil {
      target = errorAnchor
    } else if model.status == .thinking {
      target = pendingAnchor
    } else if let last = model.turns.last {
      target = last.id
    } else {
      return
    }
    withAnimation(.easeOut(duration: 0.2)) {
      proxy.scrollTo(target, anchor: .bottom)
    }
  }

  // MARK: Empty state

  private var intro: some View {
    VStack(alignment: .leading, spacing: 12) {
      Text(
        "I can see the team you have open. Ask me to fill a slot, fix a moveset, check "
          + "your coverage, or suggest a spread — edits apply to your unsaved draft, and "
          + "you keep the Save button."
      )
      .font(Theme.body(.subheadline))
      .foregroundStyle(Theme.textSecondary)

      Text("Try asking")
        .instrumentLabel()
        .foregroundStyle(Theme.textSecondary)

      FlexibleChips(items: TeamsAssistantViewModel.suggestions) { suggestion in
        Button(suggestion) { model.send(suggestion) }
          .buttonStyle(.bordered)
          .controlSize(.small)
      }
    }
    .frame(maxWidth: .infinity, alignment: .leading)
  }

  // MARK: One turn

  @ViewBuilder
  private func turnView(_ turn: TeamsAssistantViewModel.Turn) -> some View {
    VStack(alignment: .leading, spacing: 10) {
      userBubble(turn.user)
      if let answer = turn.answer {
        answerView(turn: turn, answer: answer)
      }
    }
  }

  private func userBubble(_ text: String) -> some View {
    HStack {
      Spacer(minLength: 32)
      Text(text)
        .font(Theme.body(.subheadline))
        .foregroundStyle(.white)
        .padding(.horizontal, Theme.Spacing.md)
        .padding(.vertical, Theme.Spacing.sm)
        .background(
          Theme.accent,
          in: UnevenRoundedRectangle(
            topLeadingRadius: Theme.Radius.lg,
            bottomLeadingRadius: Theme.Radius.lg,
            bottomTrailingRadius: Theme.Radius.sm,
            topTrailingRadius: Theme.Radius.lg,
            style: .continuous
          )
        )
    }
  }

  @ViewBuilder
  private func answerView(
    turn: TeamsAssistantViewModel.Turn,
    answer: BuilderAnswer
  ) -> some View {
    VStack(alignment: .leading, spacing: 10) {
      MarkdownBlockView(answer.answerMarkdown)
        .frame(maxWidth: .infinity, alignment: .leading)

      if let patch = answer.teamPatch, hasVisibleChanges(patch) {
        patchCard(turn: turn, patch: patch)
      }
    }
  }

  // MARK: Proposed-changes card

  /// Mirrors the web card's visibility rule: show only when the patch actually carries
  /// slot edits and/or a rename.
  private func hasVisibleChanges(_ patch: TeamPatch) -> Bool {
    patch.slots.count + (patch.name != nil ? 1 : 0) > 0
  }

  @ViewBuilder
  private func patchCard(
    turn: TeamsAssistantViewModel.Turn,
    patch: TeamPatch
  ) -> some View {
    VStack(alignment: .leading, spacing: 10) {
      Text("Proposed changes")
        .font(Theme.body(.subheadline).weight(.semibold))

      VStack(alignment: .leading, spacing: 6) {
        ForEach(Array(describeTeamPatch(patch).enumerated()), id: \.offset) { _, line in
          HStack(alignment: .top, spacing: 8) {
            Image(systemName: "arrow.turn.down.right")
              .font(.caption2)
              .foregroundStyle(Theme.textSecondary)
              .accessibilityHidden(true)
            Text(line)
              .font(Theme.body(.footnote))
              .fixedSize(horizontal: false, vertical: true)
          }
        }
      }

      HStack(spacing: 12) {
        if model.appliedTurnIds.contains(turn.id) {
          Label("Applied to draft", systemImage: "checkmark.circle.fill")
            .font(Theme.body(.footnote).weight(.medium))
            .foregroundStyle(Theme.success)
          if model.lastApplied?.turnId == turn.id {
            Button("Undo") { model.undo() }
              .buttonStyle(.bordered)
              .controlSize(.small)
          }
        } else {
          Button("Apply to draft") { model.apply(turn) }
            .buttonStyle(.borderedProminent)
            .controlSize(.small)
        }
      }
    }
    .padding(12)
    .frame(maxWidth: .infinity, alignment: .leading)
    .oakCard(radius: Theme.Radius.md)
  }

  // MARK: Pending / error

  private var pending: some View {
    HStack(alignment: .top, spacing: 8) {
      if model.streamingMarkdown.isEmpty {
        OakSpinner(size: 16)
        Text(model.activity ?? "Thinking\u{2026}")
          .font(Theme.body(.subheadline))
          .foregroundStyle(Theme.textSecondary)
      } else {
        MarkdownBlockView(model.streamingMarkdown)
          .frame(maxWidth: .infinity, alignment: .leading)
      }
    }
    .frame(maxWidth: .infinity, alignment: .leading)
  }

  // MARK: Composer

  private var composer: some View {
    HStack(spacing: 10) {
      TextField("Ask about this team\u{2026}", text: $input, axis: .vertical)
        .textFieldStyle(.plain)
        .lineLimit(1...4)
        .padding(.horizontal, 12)
        .padding(.vertical, 8)
        .background(Theme.surface, in: RoundedRectangle(cornerRadius: 18))
        .onSubmit(submit)
        .disabled(model.status == .thinking)

      Button(action: submit) {
        ZStack {
          Circle()
            .fill(model.canSend(input) ? Theme.accent : Theme.surfaceSunken)
            .frame(width: 38, height: 38)
          Image(systemName: "arrow.up")
            .font(Theme.body(.subheadline).weight(.semibold))
            .foregroundStyle(model.canSend(input) ? .white : Theme.textMuted)
        }
      }
      .disabled(!model.canSend(input))
      .accessibilityLabel("Send")
    }
    .padding(.horizontal, 16)
    .padding(.vertical, 10)
  }

  private func submit() {
    let text = input.trimmingCharacters(in: .whitespacesAndNewlines)
    guard model.canSend(text) else { return }
    input = ""
    model.send(text)
  }
}

// MARK: - Flexible wrapping chip row

/// A simple wrapping row of chips (the suggestion prompts). Uses a plain `HStack` with
/// wrapping via `Layout` is overkill here — three short chips fit a phone width in one
/// line, and a horizontal scroll keeps it robust for any localization.
private struct FlexibleChips<Content: View>: View {
  let items: [String]
  @ViewBuilder let content: (String) -> Content

  var body: some View {
    ScrollView(.horizontal, showsIndicators: false) {
      HStack(spacing: 8) {
        ForEach(items, id: \.self) { content($0) }
      }
    }
  }
}

File: android/app/src/main/kotlin/ai/gowtam/oak/features/teams/TeamsAssistantSheet.kt
package ai.gowtam.oak.features.teams

import ai.gowtam.oak.ui.LocalOakColors
import ai.gowtam.oak.ui.OakButton
import ai.gowtam.oak.ui.OakButtonStyle
import ai.gowtam.oak.ui.MarkdownBlockView
import ai.gowtam.oak.ui.OakRadius
import ai.gowtam.oak.ui.OakSpacing
import ai.gowtam.oak.ui.rememberHaptics
import ai.gowtam.oak.wire.TeamPatch
import ai.gowtam.oak.wire.describeTeamPatch
import androidx.compose.foundation.background
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.Send
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.WarningAmber
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.launch

/**
 * The team-builder assistant, presented from the team editor's toolbar as a
 * `ModalBottomSheet` (history-and-teams.md D-AC-TEAM2.1) — Android's stand-in for web's
 * docked `TeamsAssistantPanel.tsx` / iOS's `TeamsAssistantSheet`.
 *
 * A deliberately lean chat surface (NOT the full `AnswerCard` tree): user bubbles,
 * streamed assistant Markdown, a tool-activity ticker, and — when an answer proposes a
 * `TeamPatch` — a "Proposed changes" card with Apply/Undo. Apply mutates the editor's
 * in-memory draft only ([TeamsAssistantViewModel.apply]); the user still reviews and
 * Saves. All state lives in the injected [viewModel].
 */
@Composable
fun TeamsAssistantSheet(
    viewModel: TeamsAssistantViewModel,
    onDone: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val state by viewModel.uiState.collectAsState()
    var input by remember { mutableStateOf("") }
    val oak = LocalOakColors.current
    val haptics = rememberHaptics()

    // A mid-stream dismiss (system back, tap-outside, or Done) cancels the in-flight
    // turn and resets the status so the panel can never get stuck "thinking".
    DisposableEffect(viewModel) { onDispose { viewModel.cancel() } }

    LaunchedEffect(state.errorMessage != null) {
        if (state.errorMessage != null) haptics.error()
    }

    Column(modifier = modifier.fillMaxWidth().heightIn(min = 420.dp)) {
        Row(
            modifier = Modifier.fillMaxWidth().padding(horizontal = OakSpacing.lg, vertical = OakSpacing.md),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                "Team assistant",
                style = MaterialTheme.typography.titleMedium.copy(fontWeight = FontWeight.SemiBold),
                modifier = Modifier.semantics { heading() },
            )
            TextButton(onClick = onDone) { Text("Done") }
        }
        HorizontalDivider()

        LazyColumn(
            modifier = Modifier.weight(1f, fill = false).fillMaxWidth().padding(OakSpacing.lg),
            verticalArrangement = Arrangement.spacedBy(OakSpacing.md),
        ) {
            if (state.turns.isEmpty() && state.status != AssistantStatus.THINKING) {
                item { IntroBlock(onSuggestion = viewModel::send) }
            }
            items(state.turns, key = { it.id }) { turn -> TurnBlock(turn, state, viewModel) }
            if (state.status == AssistantStatus.THINKING) {
                item { PendingBlock(activity = state.activity, streamingMarkdown = state.streamingMarkdown) }
            }
            state.errorMessage?.let { message ->
                item {
                    ErrorRow(
                        message,
                        onRetry = if (state.errorIsRetryable) viewModel::retry else null,
                    )
                }
            }
        }
        HorizontalDivider()

        ComposerRow(
            input = input,
            onInputChange = { input = it },
            canSend = viewModel.canSend(input),
            onSend = {
                val text = input
                input = ""
                viewModel.send(text)
            },
        )
    }
}

@Composable
private fun IntroBlock(onSuggestion: (String) -> Unit) {
    val oak = LocalOakColors.current
    Column(verticalArrangement = Arrangement.spacedBy(OakSpacing.sm)) {
        Text(
            text = "I can see the team you have open. Ask me to fill a slot, fix a moveset, " +
                "check your coverage, or suggest a spread — edits apply to your unsaved draft, " +
                "and you keep the Save button.",
            style = MaterialTheme.typography.bodyMedium,
            color = oak.textMuted,
        )
        Text(
            text = "TRY ASKING",
            style = MaterialTheme.typography.labelSmall,
            color = oak.textMuted,
        )
        Row(modifier = Modifier.horizontalScroll(rememberScrollState()), horizontalArrangement = Arrangement.spacedBy(OakSpacing.sm)) {
            for (suggestion in TeamsAssistantViewModel.suggestions) {
                OakButton(onClick = { onSuggestion(suggestion) }, style = OakButtonStyle.Secondary) { Text(suggestion) }
            }
        }
    }
}

@Composable
private fun TurnBlock(turn: AssistantTurn, state: TeamsAssistantUiState, viewModel: TeamsAssistantViewModel) {
    val oak = LocalOakColors.current
    Column(verticalArrangement = Arrangement.spacedBy(OakSpacing.sm)) {
        Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.End) {
            Text(
                text = turn.user,
                style = MaterialTheme.typography.bodyMedium,
                modifier = Modifier
                    .background(oak.accent.copy(alpha = 0.14f), RoundedCornerShape(OakRadius.md))
                    .padding(horizontal = OakSpacing.md, vertical = OakSpacing.sm),
            )
        }
        turn.answer?.let { answer ->
            MarkdownBlockView(answer.answerMarkdown)
            val patch = answer.teamPatch
            if (patch != null && hasVisibleChanges(patch)) {
                val haptics = rememberHaptics()
                PatchCard(
                    patch = patch,
                    applied = turn.id in state.appliedTurnIds,
                    isLastApplied = state.lastAppliedTurnId == turn.id,
                    onApply = { haptics.success(); viewModel.apply(turn) },
                    onUndo = viewModel::undo,
                )
            }
        }
    }
}

private fun hasVisibleChanges(patch: TeamPatch): Boolean = patch.slots.isNotEmpty() || patch.name != null

@Composable
private fun PatchCard(
    patch: TeamPatch,
    applied: Boolean,
    isLastApplied: Boolean,
    onApply: () -> Unit,
    onUndo: () -> Unit,
) {
    val oak = LocalOakColors.current
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(oak.surfaceRaised, RoundedCornerShape(OakRadius.md))
            .padding(OakSpacing.md),
        verticalArrangement = Arrangement.spacedBy(OakSpacing.sm),
    ) {
        Text(
            "Proposed changes",
            style = MaterialTheme.typography.labelLarge.copy(fontWeight = FontWeight.SemiBold),
            modifier = Modifier.semantics { heading() },
        )
        Column(verticalArrangement = Arrangement.spacedBy(OakSpacing.xs)) {
            for (line in describeTeamPatch(patch)) {
                Row(horizontalArrangement = Arrangement.spacedBy(OakSpacing.xs)) {
                    Text("↳", color = oak.textMuted, style = MaterialTheme.typography.bodySmall)
                    Text(line, style = MaterialTheme.typography.bodySmall, color = oak.textStrong)
                }
            }
        }
        if (applied) {
            Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(OakSpacing.sm)) {
                Icon(Icons.Filled.CheckCircle, contentDescription = null, tint = oak.success, modifier = Modifier.size(16.dp))
                Text("Applied to draft", style = MaterialTheme.typography.labelMedium, color = oak.success)
                if (isLastApplied) {
                    OakButton(onClick = onUndo, style = OakButtonStyle.Secondary) { Text("Undo") }
                }
            }
        } else {
            OakButton(onClick = onApply) {
                Text("Apply to draft")
            }
        }
    }
}

@Composable
private fun PendingBlock(activity: String?, streamingMarkdown: String) {
    val oak = LocalOakColors.current
    if (streamingMarkdown.isEmpty()) {
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(OakSpacing.sm)) {
            CircularProgressIndicator(modifier = Modifier.size(16.dp), strokeWidth = 2.dp, color = oak.accent)
            Text(activity ?: "Thinking…", style = MaterialTheme.typography.bodyMedium, color = oak.textMuted)
        }
    } else {
        MarkdownBlockView(streamingMarkdown)
    }
}

@Composable
private fun ErrorRow(message: String, onRetry: (() -> Unit)?) {
    val oak = LocalOakColors.current
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .background(oak.warning.copy(alpha = 0.10f), RoundedCornerShape(OakRadius.md))
            .padding(OakSpacing.md),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(OakSpacing.sm),
    ) {
        Icon(Icons.Filled.WarningAmber, contentDescription = null, tint = oak.warning, modifier = Modifier.size(18.dp))
        Text(message, style = MaterialTheme.typography.bodySmall, color = oak.textStrong, modifier = Modifier.weight(1f))
        if (onRetry != null) {
            TextButton(onClick = onRetry) { Text("Retry") }
        }
    }
}

@Composable
private fun ComposerRow(input: String, onInputChange: (String) -> Unit, canSend: Boolean, onSend: () -> Unit) {
    val scope = rememberCoroutineScope()
    Row(
        modifier = Modifier.fillMaxWidth().padding(horizontal = OakSpacing.lg, vertical = OakSpacing.sm),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(OakSpacing.sm),
    ) {
        OutlinedTextField(
            value = input,
            onValueChange = onInputChange,
            placeholder = { Text("Ask about this team…") },
            modifier = Modifier.weight(1f),
            keyboardOptions = KeyboardOptions(imeAction = ImeAction.Send),
            keyboardActions = KeyboardActions(onSend = { if (canSend) scope.launch { onSend() } }),
        )
        IconButton(onClick = { if (canSend) onSend() }, enabled = canSend) {
            Icon(Icons.AutoMirrored.Filled.Send, contentDescription = "Send")
        }
    }
}
